pkg/apis/apps/validation/validation.go

```diff
@@ -129,6 +129,10 @@ func ValidateStatefulSetSpec(spec *apps.StatefulSetSpec, fldPath *field.Path, op
 
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas"))...)
 	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(spec.MinReadySeconds), fldPath.Child("minReadySeconds"))...)
+	if spec.RevisionHistoryLimit != nil {
```

Comment:
This is not allowed change, since you're breaking backwards compatibility with this kind of change. Before this change a statefulset with `.spec.revisionHistoryLimit: -1` would be accepted, but after this change you're suddenly breaking users. The correct approach for this change will be as follows (see https://kubernetes.io/docs/reference/using-api/deprecation-policy/):
1. Fix sts controller, specifically this if condition https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/pkg/controller/statefulset/stateful_set_control.go#L197 to also check for `historyLimit < 0` and `return nil` in those cases. That change can easily be backported to previous releases. Probably the best is to do that in a separate PR. 
2. Update the validation such that:
  a) in current release start with issuing a warning [OnCreate](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/pkg/registry/apps/statefulset/strategy.go#L147) and [OnUpdate](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/pkg/registry/apps/statefulset/strategy.go#L175) if this value is negative;
  b) in one of future versions start validating but only `OnCreate`, and continue issuing warnings;
  c) in one of subsequent versions start validating `OnUpdate` as well. 

You can see a similar path how we went with forbidding creating cronjob's schedule with `TZ` and `CRON_TZ` variable in https://github.com/kubernetes/kubernetes/pull/106455 and https://github.com/kubernetes/kubernetes/pull/116252. I'll defer to @liggit for the actual timeline, but that's how the process should look like, roughly.

Reply:
This change causes a crash in KCM. I'm not sure there is any cluster having this value is set to `-1` and still operates well.

Reply:
I largely agree with the comment at https://github.com/kubernetes/kubernetes/pull/129017#discussion_r1862451711

For 1, fix the crash in sts controller right away in an isolated PR. We will likely backport this fix to release branches, especially if it is tiny, well-tested, and fixes a complete controller crash.

For 2, I agree with stricter validation on create, and conditionally stricter validation on update _if_ the existing object doesn't already have a negative value. If this crashes the controller so that no object with a negative value was ever successful, we can tighten more aggressively without preceding warning-only releases, but we *do* need the conditional validation on update to handle existing persisted data with a negative value properly.

Reply:
Digging deeper, @soltysh pointed out that *technically*, the statefulset controller creates pods *before* calling `truncateHistory` the first time it processes the statefulset.

That means if a user applied a statefulset that didn't have any persistentvolumeclaim templates, they could actually get successfully running pods created *before* kube-controller-manager crashed.

For that reason, I think we should:
* fix the controller panic (already done in https://github.com/kubernetes/kubernetes/pull/129301, we can backport this)
* only warn on statefulset create or update if they specify a negative history value
    ```diff
    diff --git a/pkg/registry/apps/statefulset/strategy.go b/pkg/registry/apps/statefulset/strategy.go
    index e650001258d..ea42f4a4c02 100644
    --- a/pkg/registry/apps/statefulset/strategy.go
    +++ b/pkg/registry/apps/statefulset/strategy.go
    @@ -150,6 +151,9 @@ func (statefulSetStrategy) WarningsOnCreate(ctx context.Context, obj runtime.Obj
     	for i, pvc := range newStatefulSet.Spec.VolumeClaimTemplates {
     		warnings = append(warnings, pvcutil.GetWarningsForPersistentVolumeClaimSpec(field.NewPath("spec", "volumeClaimTemplates").Index(i), pvc.Spec)...)
     	}
    +	if newStatefulSet.Spec.RevisionHistoryLimit != nil && *newStatefulSet.Spec.RevisionHistoryLimit < 0 {
    +		warnings = append(warnings, ...)
    +	}
     	return warnings
     }
     
    @@ -182,6 +186,9 @@ func (statefulSetStrategy) WarningsOnUpdate(ctx context.Context, obj, old runtim
     	for i, pvc := range newStatefulSet.Spec.VolumeClaimTemplates {
     		warnings = append(warnings, pvcutil.GetWarningsForPersistentVolumeClaimSpec(field.NewPath("spec", "volumeClaimTemplates").Index(i).Child("Spec"), pvc.Spec)...)
     	}
    +	if newStatefulSet.Spec.RevisionHistoryLimit != nil && *newStatefulSet.Spec.RevisionHistoryLimit < 0 {
    +		warnings = append(warnings, ...)
    +	}
     
     	return warnings
     }
    ```
* allow mutating of revisionHistoryLimit on update
    ```diff
    diff --git a/pkg/apis/apps/validation/validation.go b/pkg/apis/apps/validation/validation.go
    index 7f05624fb64..c356fa5d778 100644
    --- a/pkg/apis/apps/validation/validation.go
    +++ b/pkg/apis/apps/validation/validation.go
    @@ -187,6 +187,7 @@ func ValidateStatefulSetUpdate(statefulSet, oldStatefulSet *apps.StatefulSet, op
     	newStatefulSetClone.Spec.UpdateStrategy = oldStatefulSet.Spec.UpdateStrategy             // +k8s:verify-mutation:reason=clone
     	newStatefulSetClone.Spec.MinReadySeconds = oldStatefulSet.Spec.MinReadySeconds           // +k8s:verify-mutation:reason=clone
     	newStatefulSetClone.Spec.Ordinals = oldStatefulSet.Spec.Ordinals                         // +k8s:verify-mutation:reason=clone
    +	newStatefulSetClone.Spec.RevisionHistoryLimit = oldStatefulSet.Spec.RevisionHistoryLimit // +k8s:verify-mutation:reason=clone
     
     	newStatefulSetClone.Spec.PersistentVolumeClaimRetentionPolicy = oldStatefulSet.Spec.PersistentVolumeClaimRetentionPolicy // +k8s:verify-mutation:reason=clone
     	if !apiequality.Semantic.DeepEqual(newStatefulSetClone.Spec, oldStatefulSet.Spec) {
    ```

Reply:
That will greatly simplify this PR and drop out all the ratcheting validation stuff... sorry for making you chase that path, thanks for being willing to iterate on this

Reply:
Updated based on your suggestion. Thank you for review.